### PR TITLE
Fix #174: Build failure when trying to use Dockerfile for arguments in FROM

### LINKED ISSUE #174: Build failure when trying to use Dockerfile for arguments in FROM
Looks like JKube is not able to process arguments specified in `Dockerfile`. When I use a `Dockerfile` of this format:
```
ARG VERSION:latest-java11
FROM fabric8/s2i-java:$VERSION
LABEL location="src/main/docker-context-dir"
EXPOSE 8080/tcp
COPY maven/*.jar /opt/app.jar
ENTRYPOINT [ "java", "-jar", "/opt/app.jar" ]

```
and try to build project using `mvn k8s:build -Pcontext-dir`. I'm getting this error:
```
[INFO] --- kubernetes-maven-plugin:1.0.0-SNAPSHOT:build (default-cli) @ docker-file-provided ---
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.bouncycastle.jcajce.provider.drbg.DRBG (file:/home/rohaan/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.61/bcprov-jdk15on-1.61.jar) to constructor sun.security.provider.Sun()
WARNING: Please consider reporting this to the maintainers of org.bouncycastle.jcajce.provider.drbg.DRBG
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] k8s: Running in Kubernetes mode
[INFO] k8s: Building Docker image in Kubernetes mode
[ERROR] k8s: Failed to execute the build [Error while trying to build the image]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.731 s
[INFO] Finished at: 2020-05-04T15:23:43+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.0.0-SNAPSHOT:build (default-cli) on project docker-file-provided: Failed to execute the build: Error while trying to build the image: Unable to check image [fabric8/s2i-java:$VERSION] : {"message":"no such image: fabric8/s2i-java:$VERSION: invalid reference format"} (Bad Request: 400) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
~/work/repos/jkube/quickstarts/maven/docker-file-provided : $ 
```

## Steps to reproduce:
- Go to `quickstarts/maven/docker-file-provided`
- Change `src/main/docker-file-provided/Dockerfile` like mentioned above
- Run ` mvn k8s:build -Pcontext-dir`

### PR DESCRIPTION
**Pull request by rohanKanojia**
Fix #174 

Ported PR https://github.com/fabric8io/docker-maven-plugin/pull/1299

### CODE REVIEW

**Comment by manusa** [comment]
Waiting on https://github.com/fabric8io/docker-maven-plugin/pull/1299 to be merged in order to keep changes in both projects consistent.

File: jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/helper/DockerFileUtil.java [hunk 1]
```diff
@@ -56,23 +56,40 @@ private DockerFileUtil() {}
      */
     public static List<String> extractBaseImages(File dockerFile, Properties properties) throws IOException {
         List<String[]> fromLines = extractLines(dockerFile, "FROM", properties);
+        Map<String, String> args = extractArgs(extractLines(dockerFile, "ARG", properties));
         Set<String> result = new LinkedHashSet<>();
         Set<String> fromAlias = new HashSet<>();
         for (String[] fromLine :  fromLines) {
             if (fromLine.length > 1) {
```
Comment (dev-gaur) [inline jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/helper/DockerFileUtil.java:63]: this if condition feels redundant.

Comment (dev-gaur) [reply jkube-kit/build/service/docker/src/main/java/org/eclipse/jkube/kit/build/service/docker/helper/DockerFileUtil.java:63]: Line 63

**Comment by manusa** [comment]
Why is this one closed?

**Comment by manusa** [comment]
Was this ported https://github.com/fabric8io/docker-maven-plugin/pull/1373 too? or is it not applicable?

**Comment by rohanKanojia** [comment]
ah, I still need to update this PR. Sorry, I will update it within this week

File: jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java [hunk 2]
```diff
@@ -209,4 +273,36 @@ private static File getHomeDir() {
         return new File(homeDir);
     }
 
+    private static void updateMapWithArgValue(Map<String, String> result, Map<String, String> args, String argString) {
+        if (argString.contains("=") || argString.contains(":")) {
+            String[] argStringParts = argString.split("[=:]");
+            String argStringValue = argString.substring(argStringParts[0].length() + 1);
+            if (argStringValue.startsWith("\"") || argStringValue.startsWith("'")) {
+                // Replaces surrounding quotes
+                argStringValue = argStringValue.replaceAll("^\"|\"|'|'$", "");
```
Comment (manusa) [inline jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java:282]: This Regex is dangerous, and reported as a bug https://sonarcloud.io/project/issues?id=jkubeio_jkube&open=AXVraCJxz5kaBwk-uluS&pullRequest=175&resolved=false&types=BUG

File: jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java [hunk 3]
```diff
@@ -130,6 +146,54 @@ private static Reader getFileReaderFromDir(File file) {
...
+                if (tag != null) {
+                    return imageTagArr[0] + ":" + tag;
+                }
+            }
+        }
+        return imageTagString;
+    }
+
+    static String resolveArgValueFromStrContainingArgKey(String argString, Map<String, String> args) {
+        Pattern argPattern = Pattern.compile(ARG_PATTERN_REGEX);
```
Comment (manusa) [inline jkube-kit/build/api/src/main/java/org/eclipse/jkube/kit/build/api/helper/DockerFileUtil.java:185]: Declare the pattern as a constant, it's thread-safe. This way you're compiling the pattern  for every method call.

**Review by manusa — APPROVED** [review]

**Pull request merged**